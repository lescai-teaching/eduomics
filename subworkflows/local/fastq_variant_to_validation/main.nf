include { SAMTOOLS_INDEX                 } from '../../../modules/nf-core/samtools/index/main'
include { SAMTOOLS_INDEX as INDEX_MD     } from '../../../modules/nf-core/samtools/index/main'
include { SAMTOOLS_INDEX as INDEX_RECAL  } from '../../../modules/nf-core/samtools/index/main'
include { BWA_MEM                        } from '../../../modules/nf-core/bwa/mem/main'
include { GATK4_MARKDUPLICATES           } from '../../../modules/nf-core/gatk4/markduplicates/main'
include { GATK4_BASERECALIBRATOR         } from '../../../modules/nf-core/gatk4/baserecalibrator/main'
include { GATK4_APPLYBQSR                } from '../../../modules/nf-core/gatk4/applybqsr/main'
include { GATK4_HAPLOTYPECALLER          } from '../../../modules/nf-core/gatk4/haplotypecaller/main'
include { GATK4_GENOMICSDBIMPORT         } from '../../../modules/nf-core/gatk4/genomicsdbimport/main'
include { GATK4_GENOTYPEGVCFS            } from '../../../modules/nf-core/gatk4/genotypegvcfs/main'
include { DNAVALIDATION                  } from '../../../modules/local/dnavalidation/main'


workflow FASTQ_VARIANT_TO_VALIDATION {

    take:
    reads       // channel: [mandatory] [ val(meta), [ reads ] ] <- NB: this channel contains 4 reads 2 * case + 2 * control / modified in sub-workflow
    fasta       // channel: [mandatory] [ val(meta), [ fasta ] ]
    fai         // channel: [mandatory] [ val(meta), [fai] ]
    dict        // channel: [mandatory] [ val(meta), [dict] ]
    bwa_index   // channel: [mandatory] [ val(meta), [bwa_index] ]
    dbsnp       // channel: [mandatory] [ [dbsnp] ]
    dbsnp_tbi   // channel: [mandatory] [ [dbsnp_tbi] ]
    mills       // channel: [mandatory] [ [mills] ]
    mills_tbi   // channel: [mandatory] [ [mills_tbi] ]
    capture     // channel: [mandatory] [ [capture] ]

    main:

    ch_versions = Channel.empty()

    simulated_reads_ch = reads
            .map{ m, files ->
                    def grouped = files.groupBy { file ->
                        file.name.replaceFirst(/_[12]\.fq\.gz$/, '')
                        }
                    grouped.collect { sampleName, group ->
                        def updatedMeta = m + [sample: sampleName]
                        [updatedMeta, group.sort()]
                        }
            }
            .flatMap{
                m, samplereads ->
                [m, samplereads]
            }
    simulated_reads_ch.dump(tag: 'sim reads by sample')

    // align simulated reads to their reference
    BWA_MEM( simulated_reads_ch, bwa_index, fasta, true )

    SAMTOOLS_INDEX ( BWA_MEM.out.bam )
    ch_versions = ch_versions.mix(SAMTOOLS_INDEX.out.versions)

    // ----> GATK4 MINI WORKFLOW to call normal variants <-----

    GATK4_MARKDUPLICATES(
        BWA_MEM.out.bam,
        fasta.map{ meta, it -> [ it ] },
        fai.map{ meta, it -> [ it ] },
        )
    ch_versions = ch_versions.mix(GATK4_MARKDUPLICATES.out.versions)

    empty_intervals_ch = Channel.value([[]])

    INDEX_MD(GATK4_MARKDUPLICATES.out.bam)
    ch_versions = ch_versions.mix(INDEX_MD.out.versions)

    bam_for_recal = GATK4_MARKDUPLICATES.out.bam
                        .combine(INDEX_MD.out.bai, by: 0)
                        .combine(empty_intervals_ch)
    bam_for_recal.dump(tag: 'bam for recal')

    known_sites_all = dbsnp.mix(mills).collect()
    known_sites_all_tbi = dbsnp_tbi.mix(mills_tbi).collect()

    GATK4_BASERECALIBRATOR(
        bam_for_recal,
        fasta,
        fai,
        dict,
        known_sites_all.map{ it -> [[id:'sites'], it] },
        known_sites_all_tbi.map{ it -> [[id:'sites'], it] }
        )
    ch_versions = ch_versions.mix(GATK4_BASERECALIBRATOR.out.versions)

    bam_for_applybqsr = GATK4_MARKDUPLICATES.out.bam
        .combine(INDEX_MD.out.bai, by: 0)
        .combine(GATK4_BASERECALIBRATOR.out.table, by: 0)
        .combine(capture)

    GATK4_APPLYBQSR(
        bam_for_applybqsr,
        fasta.map{ meta, it -> [ it ] },
        fai.map{ meta, it -> [ it ] },
        dict.map{ meta, it -> [ it ] }
    )
    ch_versions = ch_versions.mix(GATK4_APPLYBQSR.out.versions)

    INDEX_RECAL(GATK4_APPLYBQSR.out.bam)

    empty_models_ch = Channel.value([[]])

    bam_for_calling = GATK4_APPLYBQSR.out.bam
        .combine(INDEX_RECAL.out.bai, by: 0)
        .combine(capture)
        .combine(empty_models_ch)

    bam_for_calling.dump(tag: 'bam for calling')

    GATK4_HAPLOTYPECALLER(
        bam_for_calling,
        fasta,
        fai,
        dict,
        dbsnp.map{ it -> [[id:'test'], it] },
        dbsnp_tbi.map{ it -> [[id:'test'], it] }
    )
    ch_versions = ch_versions.mix(GATK4_HAPLOTYPECALLER.out.versions)

    // even if the samples in this execution will all be about the same variant
    // we need to collect the VCFs and we cannot just group by meta because
    // the sample ID will be different

    GATK4_HAPLOTYPECALLER.out.vcf.dump(tag: 'single sample VCF')

    gendb_input = GATK4_HAPLOTYPECALLER.out.vcf
    .join(GATK4_HAPLOTYPECALLER.out.tbi)
    .map { meta, vcf, tbi ->
        def key = meta.simulatedvar
        return [key, meta, vcf, tbi]
    }
    .groupTuple(by: 0)
    .combine(capture)
    .map { key, meta_list, vcf_list, tbi_list, interval ->
        return [meta_list[0], vcf_list, tbi_list, interval, [], []]
    }

    gendb_input.dump(tag: 'genDB channel')

    // Import GVCFs into GenomicsDB
    GATK4_GENOMICSDBIMPORT(
        gendb_input,
        false, // run_intlist
        false, // run_updatewspace
        false  // input_map
    )
    ch_versions = ch_versions.mix(GATK4_GENOMICSDBIMPORT.out.versions)

    GATK4_GENOMICSDBIMPORT.out.genomicsdb.dump(tag: 'genomicsDB out')

    // Joint genotyping
    GATK4_GENOTYPEGVCFS(
        GATK4_GENOMICSDBIMPORT.out.genomicsdb.map { meta, db -> [ meta, db, [], [], [] ] },
        fasta,
        fai,
        dict,
        dbsnp.map{ it -> [[id:'test'], it] },
        dbsnp_tbi.map{ it -> [[id:'test'], it] }
    )
    ch_versions = ch_versions.mix(GATK4_GENOTYPEGVCFS.out.versions)

    reads.dump(tag: 'original reads')

    // the sample key in a multi-sample VCF file doesn't make sense
    // and more importantly is not present in the original case/control reads
    // therefore must be removed

    validation_ch = GATK4_GENOTYPEGVCFS.out.vcf
        .map{ meta, vcf ->
            def newmeta = meta.findAll { it.key != 'sample' }
            return [newmeta, vcf]
        }
        .join(reads)

    validation_ch.dump(tag: 'validation CH')

    DNAVALIDATION( validation_ch )
    ch_versions = ch_versions.mix(DNAVALIDATION.out.versions)

    scenarios_ch = DNAVALIDATION.out.dna_validated_results
        .map { meta, results ->
            return [meta, meta.simulatedvar]
        }

    emit:
    // the following channels could be empty as a result of optional emission
    // must be handled in main before they are passed to the aiscenarios module
    simulation = DNAVALIDATION.out.dna_validated_results  // channel: [ val(meta), path(validated_results_folder/*) ]
    scenario   = scenarios_ch                             // channel: [ val(meta), val(variant) ]
    versions   = ch_versions                              // channel: [ versions.yml ]

}
