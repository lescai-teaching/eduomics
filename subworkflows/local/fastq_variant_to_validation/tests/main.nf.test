nextflow_workflow {

    name "Test Subworkflow FASTQ_VARIANT_TO_VALIDATION"
    script "../main.nf"
    workflow "FASTQ_VARIANT_TO_VALIDATION"

    config "./test.config"
    options "-dump-channels"

    tag "subworkflows"
    tag "subworkflows_"
    tag "subworkflows/fastq_variant_to_validation"

    tag "samtools"
    tag "samtools/faidx"
    tag "samtools/index"
    tag "samtools/sort"
    tag "bwa/index"
    tag "bwa/mem"
    tag "gatk4"
    tag "gatk4/markduplicates"
    tag "gatk4/baserecalibrator"
    tag "gatk4/createsequencedictionary"
    tag "gatk4/applybqsr"
    tag "gatk4/haplotypecaller"

    // TODO nf-core: Change the test name preferably indicating the test-data and file-format used
    test("single sample - variant present") {

        when {
            workflow {
                """
                // channel: [mandatory] [ val(meta), [ reads ] ] <- NB: this channel contains 4 reads 2 * case + 2 * control / modified in sub-workflow
                input[0] = Channel.of([
                    [ id: 'sim01', type: 'dnaseq', chromosome: 'chr22', reps: 'none', groups: 2, coverage: 50, simulatedvar: 'chr22-18078510-T-C' ], // meta map
                    [
                        file('https://raw.githubusercontent.com/lescai-teaching/eduomics_testdata/refs/heads/main/dna/simureads/chr22-18078510-T-C/normal_0.000+disease_1.000_1.fq.gz', checkIfExists: true),
                        file('https://raw.githubusercontent.com/lescai-teaching/eduomics_testdata/refs/heads/main/dna/simureads/chr22-18078510-T-C/normal_0.000+disease_1.000_2.fq.gz', checkIfExists: true),
                        file('https://raw.githubusercontent.com/lescai-teaching/eduomics_testdata/refs/heads/main/dna/simureads/chr22-18078510-T-C/normal_1.000+disease_0.000_1.fq.gz', checkIfExists: true),
                        file('https://raw.githubusercontent.com/lescai-teaching/eduomics_testdata/refs/heads/main/dna/simureads/chr22-18078510-T-C/normal_1.000+disease_0.000_2.fq.gz', checkIfExists: true)
                    ]
                ])

                // channel: [mandatory] [ val(meta), [ fasta ] ]
                input[1] = Channel.value([
                    [ id:'test', coverage:50 ], // meta map
                    file('https://raw.githubusercontent.com/lescai-teaching/eduomics_testdata/refs/heads/main/dna/whole_chr22/Homo_sapiens.GRCh38.dna.chromosome.22.fa', checkIfExists: true),
                ])

                // channel: [mandatory] [ val(meta), [fai] ]
                input[2] = Channel.value([
                    [ id:'test', coverage:50 ], // meta map
                    file('https://raw.githubusercontent.com/lescai-teaching/eduomics_testdata/refs/heads/main/dna/whole_chr22/Homo_sapiens.GRCh38.dna.chromosome.22.fa.fai', checkIfExists: true),
                ])

                // channel: [mandatory] [ val(meta), [dict] ]
                input[3] = Channel.value([
                    [ id:'test', coverage:50 ], // meta map
                    file('https://raw.githubusercontent.com/lescai-teaching/eduomics_testdata/refs/heads/main/dna/whole_chr22/Homo_sapiens.GRCh38.dna.chromosome.22.dict', checkIfExists: true),
                ])

                // channel: [mandatory] [ val(meta), [bwa_index] ]
                input[4] = Channel.value([
                    [ id:'test', coverage:50 ], // meta map
                    [
                    file('https://raw.githubusercontent.com/lescai-teaching/eduomics_testdata/refs/heads/main/dna/whole_chr22/whole_22_bwa_index/Homo_sapiens.GRCh38.dna.chromosome.22.fa.amb', checkIfExists: true),
                    file('https://raw.githubusercontent.com/lescai-teaching/eduomics_testdata/refs/heads/main/dna/whole_chr22/whole_22_bwa_index/Homo_sapiens.GRCh38.dna.chromosome.22.fa.ann', checkIfExists: true),
                    file('https://raw.githubusercontent.com/lescai-teaching/eduomics_testdata/refs/heads/main/dna/whole_chr22/whole_22_bwa_index/Homo_sapiens.GRCh38.dna.chromosome.22.fa.bwt', checkIfExists: true),
                    file('https://raw.githubusercontent.com/lescai-teaching/eduomics_testdata/refs/heads/main/dna/whole_chr22/whole_22_bwa_index/Homo_sapiens.GRCh38.dna.chromosome.22.fa.fai', checkIfExists: true),
                    file('https://raw.githubusercontent.com/lescai-teaching/eduomics_testdata/refs/heads/main/dna/whole_chr22/whole_22_bwa_index/Homo_sapiens.GRCh38.dna.chromosome.22.fa.pac', checkIfExists: true),
                    file('https://raw.githubusercontent.com/lescai-teaching/eduomics_testdata/refs/heads/main/dna/whole_chr22/whole_22_bwa_index/Homo_sapiens.GRCh38.dna.chromosome.22.fa.sa', checkIfExists: true),
                    ]
                ])

                // channel: [mandatory] [ [dbsnp] ]
                input[5] = Channel.value([
                    file('https://raw.githubusercontent.com/lescai-teaching/eduomics_testdata/refs/heads/main/dna/whole_chr22/dbsnp_146.hg38_chr22.vcf.gz', checkIfExists: true),
                ])

                // channel: [mandatory] [ [dbsnp_tbi] ]
                input[6] = Channel.value([
                    file('https://raw.githubusercontent.com/lescai-teaching/eduomics_testdata/refs/heads/main/dna/whole_chr22/dbsnp_146.hg38_chr22.vcf.gz.tbi', checkIfExists: true),
                ])

                // channel: [mandatory] [ [mills] ]
                input[7] = Channel.value([
                    file('https://raw.githubusercontent.com/lescai-teaching/eduomics_testdata/refs/heads/main/dna/whole_chr22/Mills_and_1000G_gold_standard.indels.hg38_chr22.vcf.gz', checkIfExists: true),
                ])

                // channel: [mandatory] [ [mills_tbi] ]
                input[8] = Channel.value([
                    file('https://raw.githubusercontent.com/lescai-teaching/eduomics_testdata/refs/heads/main/dna/whole_chr22/Mills_and_1000G_gold_standard.indels.hg38_chr22.vcf.gz.tbi', checkIfExists: true),
                ])

                // channel: [mandatory] [ [capture] ]
                input[9] = Channel.value([
                    file('https://raw.githubusercontent.com/lescai-teaching/eduomics_testdata/refs/heads/main/dna/whole_chr22/Twist_exome_2.0_covered_chr22_500pad.bed', checkIfExists: true),
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }
}
