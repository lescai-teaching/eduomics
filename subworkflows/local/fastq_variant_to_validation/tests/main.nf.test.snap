{
    "single sample - variant present": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "2": [
                    "versions.yml:md5,2250f3ebc572a41c75435e2f607597c9",
                    "versions.yml:md5,2250f3ebc572a41c75435e2f607597c9",
                    "versions.yml:md5,31239f47332a0c0c0487e58e6d4340d1",
                    "versions.yml:md5,31239f47332a0c0c0487e58e6d4340d1",
                    "versions.yml:md5,49b101185b9cf7dc6867351331fcdece",
                    "versions.yml:md5,870aef698dcae3d52b4e97fb42d5dc3c",
                    "versions.yml:md5,a4ded6bc9584272681d0aa8147e0d83e",
                    "versions.yml:md5,a4ded6bc9584272681d0aa8147e0d83e",
                    "versions.yml:md5,be3b8ea9e6bcec95af940e3aed132b18",
                    "versions.yml:md5,f09aefdd9471bac92294842e2e656e3a",
                    "versions.yml:md5,f09aefdd9471bac92294842e2e656e3a",
                    "versions.yml:md5,f0a2ce5fd9c9e5eef282404f98543270",
                    "versions.yml:md5,f774466b3de5cf34f0523d6493f5d82a",
                    "versions.yml:md5,f774466b3de5cf34f0523d6493f5d82a"
                ],
                "scenario": [
                    
                ],
                "simulation": [
                    
                ],
                "versions": [
                    "versions.yml:md5,2250f3ebc572a41c75435e2f607597c9",
                    "versions.yml:md5,2250f3ebc572a41c75435e2f607597c9",
                    "versions.yml:md5,31239f47332a0c0c0487e58e6d4340d1",
                    "versions.yml:md5,31239f47332a0c0c0487e58e6d4340d1",
                    "versions.yml:md5,49b101185b9cf7dc6867351331fcdece",
                    "versions.yml:md5,870aef698dcae3d52b4e97fb42d5dc3c",
                    "versions.yml:md5,a4ded6bc9584272681d0aa8147e0d83e",
                    "versions.yml:md5,a4ded6bc9584272681d0aa8147e0d83e",
                    "versions.yml:md5,be3b8ea9e6bcec95af940e3aed132b18",
                    "versions.yml:md5,f09aefdd9471bac92294842e2e656e3a",
                    "versions.yml:md5,f09aefdd9471bac92294842e2e656e3a",
                    "versions.yml:md5,f0a2ce5fd9c9e5eef282404f98543270",
                    "versions.yml:md5,f774466b3de5cf34f0523d6493f5d82a",
                    "versions.yml:md5,f774466b3de5cf34f0523d6493f5d82a"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-05-21T17:02:20.126393124"
    }
}